shell/mixins/harvester-vm/index.js

```diff
@@ -697,44 +698,39 @@ export default {
 
     getInitUserData(config) {
       const _QGA_JSON = this.getMatchQGA(config.osType);
+
       const out = jsyaml.dump(_QGA_JSON);
 
-      return out.startsWith('#cloud-config') ? out : `#cloud-config\n${ out }`;
+      return `#cloud-config\n${ out }`;
     },
 
     getUserData(config) {
-      const { returnType = 'string' } = config;
-
-      let userDataJson = this.convertToJson(this.userScript) || {};
-
-      const sshAuthorizedKeys = userDataJson?.ssh_authorized_keys || [];
-
-      if (userDataJson && sshAuthorizedKeys) {
-        const sshList = new Set([...this.getSSHListValue(this.sshKey), ...sshAuthorizedKeys]);
-
-        userDataJson.ssh_authorized_keys = [...sshList];
-      } else {
-        userDataJson.ssh_authorized_keys = this.getSSHListValue(this.sshKey);
-      }
-
-      if (userDataJson.ssh_authorized_keys && userDataJson.ssh_authorized_keys.length === 0) {
-        delete userDataJson.ssh_authorized_keys;
-      }
+      try {
+        if (!this.userScript) {
+          return undefined;
+        }
+        let userDataDoc = YAML.parseDocument(this.userScript);
 
-      userDataJson = config.installAgent ? this.mergeQGA({ userDataJson: clone(userDataJson), ...config }) : this.deleteQGA({ userDataJson, ...config });
+        const allSSHAuthorizedKeys = this.mergeSSHAuthorizedKeys(this.userScript);
 
-      if (returnType === 'string') {
-        const out = jsyaml.dump(userDataJson);
+        if (allSSHAuthorizedKeys.length > 0) {
+          userDataDoc.setIn(['ssh_authorized_keys'], allSSHAuthorizedKeys);
+        } else if (YAML.isCollection(userDataDoc.getIn('ssh_authorized_keys'))) {
+          userDataDoc.deleteIn(['ssh_authorized_keys']);
+        }
 
-        const outValue = out.replace(/[\r\n]/g, '').replace(/\ +/g, '');
+        userDataDoc = config.installAgent ? this.mergeQGA({ userDataDoc, ...config }) : this.deleteQGA({ userDataDoc, ...config });
+        const userDataYaml = userDataDoc.toString();
 
-        if (outValue === "''") {
-          return undefined;
+        if (userDataYaml === '{}\n') {
+          return `#cloud-config`;
```

Comment:
When the user unchecks the `install-agent`, we want the line `#cloud-config` not to be removed, because the "user data" passed into the VM will only take effect if `#cloud-config` is present.

Reply:
But is it an anchor of an URL or just some Harvester configuration?

Reply:
Yep, Just some Harvester YAML configuration that is used by the created virtual machine.

Reply:
Ok got it, thank you so much for the patience!
I think it may be worth to note it down in the code, so we dont waste the effort 😄

Reply:
Thx, I will add the comment to the code.